pkg/providers/secretmanager/secretmanager.go

```diff
@@ -0,0 +1,87 @@
+package secretmanager
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+
+	smpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1beta1"
+
+	sm "cloud.google.com/go/secretmanager/apiv1beta1"
+	"github.com/variantdev/vals/pkg/api"
+)
+
+// Format: ref+secretmanager://project/mykey[?version=VERSION]#/yaml_or_json_key/in/secret
+type provider struct {
+	client  *sm.Client
+	ctx     context.Context
+	version string
+}
+
+func New(cfg api.StaticConfig) *provider {
+	ctx := context.Background()
+
+	p := &provider{
+		ctx: ctx,
+	}
+
+	version := cfg.String("version")
+	if version == "" {
+		p.version = "latest"
+	} else {
+		p.version = version
+	}
+
+	return p
+}
+
+func (p *provider) GetString(key string) (string, error) {
+
+	secret, err := p.getSecretBytes(key)
+	if err != nil {
+		return "", err
+	}
+
+	return string(secret), nil
+}
+
+func (p *provider) GetStringMap(key string) (map[string]interface{}, error) {
+
+	secretMap := map[string]interface{}{}
+
+	secretString, err := p.GetString(key)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := yaml.Unmarshal([]byte(secretString), secretMap); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal secret: %w", err)
+	}
+
+	return secretMap, nil
+}
+
+func (p *provider) getSecretBytes(key string) ([]byte, error) {
+
+	c, err := sm.NewClient(p.ctx)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to connect: %s", err)
+		return nil, err
+	}
+	splitKey := strings.SplitN(key, "/", 2)
```

Comment:
Does GCP Secrets Manager allow omitting the version number?

If required, I believe we the key format should be `ref+gcpsecrets://PROJECT/SECRET/VERSION`, without `?version=VERSION`. I've been using the query parameter only for something optional.

You could also validate the length of `splitKey` then - like:

```
if len(splitKey) != 3 {
    return fmt.Errorf(`invalid key %q: required format "PROJECT/SECRET/VERSION"`, key)
}
```

Reply:
Just saw this https://github.com/variantdev/vals/pull/20/files#diff-e0b710683a5ede707fa1a9d0643dcc1aR33

Would `latest` the well-known default version that should always be available? Then I think `VERSION` can be said as optional. In that case I'm fine with `version` in the query rather the URL path.

Reply:
>Would latest the well-known default version that should always be available?

Yes it is always the newest (or only) one.